solidity/contracts/CronosGravity.sol

```diff
@@ -70,17 +80,56 @@ contract CronosGravity is Gravity, AccessControl, Pausable, Ownable {
         _unpause();
     }
 
+    /**
+    * Only owner
+    * Notify the start of the migration
+    */
+    function startMigration() public onlyOwner {
+        migrationHeight = block.number + migrationPeriod;
+        migration = true;
+        emit Migration(true);
+    }
+
+    /**
+    * Only owner
+    * Stop the migration
+    */
+    function stopMigration() public onlyOwner {
+        // add approximately one day to the migration start height
+        migrationHeight = 0;
+        migration = false;
+        emit Migration(false);
+    }
+
+    /**
+    * Only owner
+    * Set a new migration period
+    */
+    function setMigrationPeriod(uint256 _migrationPeriod) public onlyOwner {
+        require(_migrationPeriod >= 6500, "Migration period cannot be lower than 6500");
+        migrationPeriod = _migrationPeriod;
+    }
+
     /**
     * Only owner
     * migrateToken allows to migrate locked fund to a new gravity contract
     * in case we need to upgrade it
     */
     function migrateToken(
         address _tokenContract,
-        address _destination,
-        uint256 _amount
+        address _newGravityAddress,
+        uint256 _amount,
+        bool isCosmosToken
     ) public onlyOwner {
-        IERC20(_tokenContract).safeTransfer(_destination, _amount);
+        require(migration == true, "Migration has not started");
+        require(migrationHeight != 0, "Migration height cannot be zero");
+        require(block.number >= migrationHeight, "Migration is not allowed yet");
+        require(block.number <= migrationHeight + migrationPeriod, "Migration time has exceeded");
+        if (isCosmosToken) {
+            ICosmosToken(_tokenContract).setGravityContract(_newGravityAddress);
```

Comment:
The function name setGravityContract is misleading as it is performing transfer. Can we change it to transferAndSetGravityContract

Reply:
unfortunately, the interface cannot be changed anymore since we have (or rather soon) deployed cosmoserc20 token

Reply:
Looks good to me. The whole migration logic can be started after 3 days of setting migrationHeight, and the migration operation will last for 3 days or controlled(ended) by owner during the migration period.

Reply:
yes, we detect that the migration has started unintentionally, we also have 3 days to react and stop the migration before it happening